src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7164,12 +7164,54 @@ bool canFollowNullableType()
                     case ParseTypeMode.AfterIs:
                     case ParseTypeMode.DefinitePattern:
                     case ParseTypeMode.AsExpression:
-                        // These contexts might be a type that is at the end of an expression. In these contexts we only
-                        // permit the nullable qualifier if it is followed by a token that could not start an
-                        // expression, because for backward compatibility we want to consider a `?` token as part of the
-                        // `?:` operator if possible.
-                        //
-                        // Similarly, if we have `T?[]` we do want to treat that as an array of nullables (following
+                        // These contexts might be a type that is at the end of an expression.
+                        // For backward compatibility we want to consider a `?` token as part of the `?:` operator if possible.
+                        // However, if current token is an identifier, it might be beneficial to allow `?`
+                        // and treat this identifier as a designation. Nullable types in patterns are semantically invalid,
+                        // so we will get a nice error about that during binding
+                        if (this.CurrentToken.Kind == SyntaxKind.IdentifierToken)
+                        {
+                            // Given that we are deciding whether we take `?:` operator or not while looking at state after `?`,
+                            // our focus is on `:` token as well. But `:` have special meaning in context of switch statements
+                            // and expressions (since `:` in switch expressions is recovered to `=>`). So consider the following case:
+                            // `case T? t:`. On its own `T ? t :` can be treated as a conditional expression with condition `T`,
+                            // 'whenTrue' `t` and 'whenFalse' not yet parsed. But it is way better to take `:` as case label end and
+                            // thus parse label condition as a declaration pattern `T? t`. `whenIsKeyword` flag can conveniently tell
+                            // whether we are in a 'top-level' state of a label, so if condition get parenthesized like `case (a ? b : c):`
+                            // we don't take the first `:` as label end and parse label condition as a parenthesized conditional expression
+                            if ((_termState.HasFlag(TerminatorState.IsExpressionOrPatternInCaseLabelOfSwitchStatement) ||
```

Comment:
i really dont' like termstate checks.  they look up arbitrarily high in the parser stack, and don't necessarily mean the code we're *currently* parsing has anything to do with that.  we could also just be in something like a simple nested parenthesized expression that's nowhere neear the end of the case label.

For this, if you want to support this sort of flexible control, a boolean should be passed through indicating the current context.  Note: i would prefer the boolean *not* be optional so it's clear which callers are impacted and what they are passing in.

Note: if we do this sort of _termState examination *anywherE* other than the code that is trying to resync, that's bad imo.

Reply:
> a boolean should be passed through indicating the current context

This is basically `whenIsKeyword` flag. `when` can only be a keyword in top-level context of a case lable/switch expression pattern. There are tests like `case (T? when) when x > 0`, which verify that code path (i.e. removing `whenIsKeyword` flag makes them fail)

I mentined that in the comment above that `if`. Maybe you can suggest a version of that comment, which is more clear bout the intent?

Reply:
nothing changes about my opinoin of _termpState.  We should not be checking it.  It is ambient state about information *arbitrarily* high up in teh parsing stck.

Reply:
Stopping review at this point.  We need another way to not use this.